packages/action-menu/README.md

```diff
@@ -50,28 +48,26 @@ The visible label that is be provided via the default `<slot>` interface can be
 
 <!-- prettier-ignore -->
 ```html
-<sp-action-menu label="More actions">
-    <sp-menu slot="options" style="min-width: 125px">
-        <sp-menu-item>
-            Deselect
-        </sp-menu-item>
-        <sp-menu-item>
-            Select inverse
-        </sp-menu-item>
-        <sp-menu-item>
-            Feather...
-        </sp-menu-item>
-        <sp-menu-item>
-            Select and mask...
-        </sp-menu-item>
-        <sp-menu-divider></sp-menu-divider>
-        <sp-menu-item>
-            Save selection
-        </sp-menu-item>
-        <sp-menu-item disabled>
-            Make work path
-        </sp-menu-item>
-    </sp-menu>
+<sp-action-menu label="More Actions" menu-width="125">
```

Comment:
It feels a bit weird having presentation mixed with structure here. I feels like we should be able to set a CSS custom property. However, I realise that's a problem with the overlay system. I wonder if components need to be able to specify custom properties that the overlay system should copy over to the overlay?

Reply:
We _should_ be able to grab this value from CSS Custom Properties, the only issue is that there is no lifecycle management of those values off of which to hang change management. Do you have thoughts on a contract that you'd be OK with keeping here? Here's a could that I can think of offhand:

- check on the `firstUpdate` of `sp-dropdown`, only (super restrictive, probably always early)
- check on every `update` of `sp-dropdown` (may become performance prohibitive)
- check on every `menuOpen` (probably my suggestion, but a difficult call)
- add an imperative API for an implementor to call when changing (awkward...)
- `resizeObserver` (not hot on the idea for x-browser purposes, but might be the most "correct")

The second question is what do we target for setting the width. The `sp-dropdown` relies on the `--spectrum-global-dimension-size-2400` custom property (which seems like a Spectrum CSS issue) for its `width` and `--spectrum-dropdown-min-width` for its `min-width`, `sp-popover` relies on `--spectrum-global-dimension-size-400` for it's `min-width`, while `sp-menu` and `sp-menu-item` don't seem to have `width` related custom properties at this point. We _could_ pull the `width` value straight out of `sp-menu-item` when we query through them so as to allow a more direct sizing API even if the implementor isn't actually sizing the right `sp-menu-item` it would feel like they are.

Thoughts?

Reply:
I can see why you had side-stepped this one 😄 

I am happy with "check on every menuOpen." That seems like the right granularity. If it isn't, I presume we will hear in the form of new issues.

So, we definitely don't want to make users override `--spectrum-global-dimension-size-2400`. That does sound like an issue with spectrum-css. Maybe we need to define `--spectrum-dropdown-width: var(--spectrum-global-dimension-size-2400);`?

Since CSS custom props are going to be a part of our exposed API, we should file issues with spectrum-css when we hit these. In the meantime, I guess we try to tidy this stuff up in our CSS overrides.

Intuitively, it makes sense to me to set min/max and let the contained items influence the size within bounds. However, I don't know how that sits with the designs. We should probably stick with the spectrum-css implementation for now and refine if needed??

Reply:
I'm not sure I see what you're saying about an API that relies primarily on min/max as opposed to simple width. If you have strong feelings there, I'd love to hear more of what you see that covering/supporting to see if I can get on board with it.

However, thinking this through for myself, it's not really `--spectrum-dropdown-width` that we want here (though I think that should probably be added upstream), as that should manage the width of the element visible at all times, the "button". That element should have the ability to be delivered at a different with than the actual "menu" element. To that end, I added `--spectrum-dropdown-menu-width`, and started passing it into the `sp-popover` element on each call to `openMenu()`. The code flow there makes me even more reticent to address the possibility of an API to pass more complex custom property values in the `Overlay` system at large, but I think it works while having some performance safety built-in.

Should this be more thoroughly documented? We've been punting a bit of the CSS Custom Property documentation system-wide, but maybe this _one_ deserves closer work? It's possible this is something we come back to?

Reply:
Is it possible to avoid this copying of the custom property entirely? Should content in the dropdown not follow its own natural width, and potentially allow us to have the dropdown pass on its computed width to the component to force that the dropdown content matches the width of the control? I think both cases are called for?

Reply:
While I think that we'd want to learn from the failings of the native `select` element and surface the ability to control the width of the overlaid menu itself at some point, this is certainly a possibility for simplifying this early iteration. Thinking on this again, it also points to subtle support of the quiet variant that seems to be missing currently.

Particularly, it does look like the general expectation is for the [standard variant](https://opensource.adobe.com/spectrum-css/components/dropdown/#Standard) to fill the full width of the field button:
![image](https://user-images.githubusercontent.com/1156657/74286676-a008eb80-4d6b-11ea-83ab-54d4bde5e3ed.png)
While the [quiet variant](https://opensource.adobe.com/spectrum-css/components/dropdown/#Quiet) allows for the content to decide for itself the width it'll display:
![image](https://user-images.githubusercontent.com/1156657/74286728-c4fd5e80-4d6b-11ea-8e49-12f3d97b20e2.png)

All three realities can be supported by a small change to the current code, a la:
```
       const menuWidthCustomPropertyValue = this.style.getPropertyValue(
            '--spectrum-dropdown-menu-width'
        );
        const hasCustomProperty =
            menuWidthCustomPropertyValue.search('-') !== -1;
        // only use `getComputedStyle(this)` if it's a custom property itself
        const computedMenuWidth = hasCustomProperty
            ? getComputedStyle(this).getPropertyValue(
                  '--spectrum-dropdown-menu-width'
              )
            : menuWidthCustomPropertyValue;
        // only use `this.offsetWidth` when no value is available && Standard varient
        const menuWidth = computedMenuWidth
            ? computedMenuWidth
            : !this.quiet && `${this.offsetWidth}px`;
        if (menuWidth) {
            this.popover.style.setProperty(
                '--spectrum-dropdown-menu-width',
                menuWidth
            );
        }
```
Future versions could even look into caching to possibly reduce repeated measuring, but locking down the expected API seems much more important at this point.

Or we could use a simplified (either in this sort of structure or directly in the `style` tag):
```
        // only use `this.offsetWidth` when Standard varient
        const menuWidth = !this.quiet && `${this.offsetWidth}px`;
        if (menuWidth) {
            this.popover.style.setProperty(
                '--spectrum-dropdown-menu-width',
                menuWidth
            );
        }
```
And then revisit the ability to deliver more granular customization of the visual delivery later.